File: Sources/windows/yage/capabilities/auto_respawn.py
import random
from typing import List
from yage.models.capability import Capability
from yage.models.collisions import Collision
from yage.models.entity_state import EntityState
from yage.models.hotspots import BOUNDS_THICKNESS
from yage.utils.geometry import Point, Rect, Vector
from yage.utils.logger import Logger


class AutoRespawn(Capability):
    def do_update(self, collisions: List[Collision], time: float):
        if self.subject.state == EntityState.DRAG:
            return
        if self.is_within_bounds(self.subject.frame.origin):
            return
        Logger.log(self.tag, "Outside of bounds, teleporting...")
        self.teleport()

    def teleport(self):
        world_width = self.subject.world_bounds.width
        random_x = world_width * random.uniform(0.2, 0.8)
        self.subject.frame.origin = Point(x=random_x, y=30)
        self.subject.direction = Vector(dx=1, dy=0)
        self.subject.state = EntityState.MOVE

    def outer_bounds(self):
        bounds = self.subject.world.bounds if self.subject else Rect.zero()
        bounds = bounds.offset(x=-bounds.min_x, y=-bounds.min_y)
        bounds = bounds.inset_by(-BOUNDS_THICKNESS * 5)
        return bounds

    def is_within_bounds(self, point: Point) -> bool:
        return self.outer_bounds().contains(point)

File: Sources/macOS/Sources/Yage/Capabilities/AutoRespawn.swift
import Combine
import Schwifty
import SwiftUI

public class AutoRespawn: Capability {
    override public func doUpdate(with collisions: Collisions, after time: TimeInterval) {
        guard let subject, subject.state != .drag else { return }
        if !isWithinBounds(point: subject.frame.origin) {
            Logger.log(tag, subject.id, "Teleporting...")
            teleport()
        }
    }

    public func teleport() {
        guard let subject else { return }
        let worldWidth = subject.worldBounds.width
        let randomX = worldWidth * CGFloat.random(in: 0.05...0.95)
        subject.frame.origin = CGPoint(x: randomX, y: 30)
        subject.direction = CGVector(dx: 1, dy: 0)
        subject.set(state: .move)
    }

    func outerBounds() -> CGRect {
        (subject?.worldBounds.bounds ?? .zero).inset(by: -boundsThickness*5)
    }

    func isWithinBounds(point: CGPoint) -> Bool {
        let bounds = outerBounds()
        if bounds.contains(point) { return true }
        Logger.log(tag, subject?.id ?? "", "at", point.description, "is outside", bounds.description)
        return false
    }
}
